src/bpfprogram.cpp

```diff
@@ -0,0 +1,77 @@
+#include "bpfprogram.h"
+
+#include <optional>
+#include <stdexcept>
+#include <tuple>
+
+namespace bpftrace {
+
+std::optional<BpfProgram> BpfProgram::CreateFromBytecode(
+    const BpfBytecode &bytecode,
+    const char *name,
+    MapManager &maps)
+{
+  if (bytecode.find(name) != bytecode.end())
+  {
+    return BpfProgram(bytecode, name, maps);
+  }
+  return std::nullopt;
+}
+
+BpfProgram::BpfProgram(const BpfBytecode &bytecode,
+                       const char *name,
+                       MapManager &maps)
+    : bytecode_(bytecode), maps_(maps), name_(name), code_()
+{
+}
+
+const std::vector<uint8_t> &BpfProgram::getCode()
+{
+  if (code_.empty())
```

Comment:
Not wrong specifically, but the flow of the code through `getCode` and `assemble` is a little odd and seems like a recipe for bugs
- why do we check `code_.empty()` so many times?
- right now we never call getCode() without first calling assembel(), which seems reasonable. This invariant isn't defined anywhere and this block of code I'm commenting on would hide violations of this
- we're not catching the exceptions in this code path

As it seems to work and in the interests of moving forward with BPF subprograms, it probably doesn't have to be changed right away though

Reply:
This is the original design that I kept but I agree here that we should be always explicitly calling `assemble` before we work with the BPF program, otherwise it's a bug. So let me remove this check.